### PR TITLE
Flax nnx

### PR DESCRIPTION
**Pull request by sachaguer**
A PR to modify the code from Flax Linen to Flax NNX. Flax NNX is more intuitive to use and more flexible to add neural network architecture with Batch Normalization, Dropout or any stochastic layer. Flax Linen requires to rewrite the training loops to account for those stochastic layers whereas Flax NNX does not. These are the steps to build this new API:
- Modify the Neural Network architecture with NNX (compressors and normalizing flows).
- Modify the training loop.
- Modify the methods to load and save checkpoints
- Modify the unit tests and pass them.

### CODE REVIEW

**Comment by sachaguer** [comment]
- Loss functions also have to be modified as the params are no longer an argument of the loss, only the model (carrying the parameters in its attribute)

**Comment by sachaguer** [comment]
PR is ready. All the unit tests passed. Only thing to modify is the examples to make sure they run with the new nnx interface.
